### ISSUE TITLE
Support logical assignments

### ISSUE THREAD

**Issue by mdjermanovic** [eslint/eslint#13569]
Logical assignment proposal is now Stage 4.

https://github.com/tc39/proposal-logical-assignment

ESLint's default parser will support this syntax under `ecmaVersion: 2021`

- [x] **ESTree:** https://github.com/estree/estree/pull/233
- [x] **Acorn:** https://github.com/acornjs/acorn/pull/968
- [x] **Espree:** https://github.com/eslint/espree/pull/448
- [x] **ESLint:**
  - [x] Code path analysis #13612
  - [x] `operator-assignment` #13618
  - [x] `constructor-super` #13618
  - [x] `astUtils.couldBeError` #13618
  - [x] `no-constant-condition` #13946
  - [x] `complexity` #13979
  - [x] `func-names` #14908  
  - [x] `no-self-assign`? #14152

**Comment by mdjermanovic** [comment]
These rules/helpers should be probably updated:

* `operator-assignment`: Currently, `"never"` reports `a ||= b` (and auto-fixes some cases that shouldn't be auto-fixed), but `"always"` doesn't report `a = a || b`, which is inconsistent behavior. The same applies to `&&=` and `??=`. Possible solutions:
  * Fix the fixer for `"never"`, and also check logical expressions on the right side (like `a = a || b`) if the option is `"always"` and `ecmaVersion >= 2021`. Though, we're usually trying to avoid `ecmaVersion` in rules. Also, update the documentation because it explicitly lists operators.
  * Or, fix the rule to just ignore `a ||= b` and other logical assignments.
  * Or, add a new option for logical assignments.
* `constructor-super`: Since the rule already allows `super()` in constructor of `class A extends (B || null)`, then it should probably allow `super()` in constructor of `class A extends (B ||= null)`.
* `astUtils.couldBeError`: returns `true` for  `a || 5`, so it should probably return `true` for `a ||= 5`, too.

It looks like these rules should be also updated, but it might be a breaking change for babel-eslint users:

* `complexity`: I guess `||=` increases complexity just like `||`. Same for others.
* `no-constant-condition`: This rule already reports `a = true` and `a || true`, so it should probably report `a ||= true` as well. Same for `a &&= false`.
* `no-self-assign`: Should it report `a ||= a`, `a &&= a`, and `a ??= a`?

This rule probably needs a documentation update:

* `func-names`, option `"as-needed"`:  Per the [specification](https://tc39.es/ecma262/#sec-assignment-operators-runtime-semantics-evaluation) it seems that all three of `||=`, `&&=`, and `??=` do assign function names (`NamedEvaluation`). The rule works in line with that fact, because it doesn't check the assignment operator (and also mistakenly allows `+=` and other operators). However, we should probably rephrase the [description in the docs](https://eslint.org/docs/rules/func-names#options-29) since the condition `"if the name cannot be assigned automatically in an ES6 environment"` sounds ambiguous now.

These rules probably shouldn't be updated:

* Rules that account for logical operands, but don't check assignments, like `array-callback-return`, `prefer-arrow-callback`, `no-return-await`, etc.

Some other rules and helpers not listed here don't check `operator` in assignment expressions, it might be good to doublecheck them.

**Comment by mdjermanovic** [comment]
Marked as "breaking" to update the `complexity` rule in the next major version.

**Comment by Sembiance** [comment]
Node.js v15 now supports logitcal assignment operators: https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V15.md

Chrome 85, Firefox 79 and Safari 14 all support it too.

With Node now supporting it, more and more developers are going to be adopting it I'm sure.

**Comment by mdjermanovic** [comment]
ESLint also supports logical assignment operators, as of [v7.8.0](https://eslint.org/blog/2020/08/eslint-v7.8.0-released).

If you're using the default parser, just update the `ecmaVersion` to the latest `2021`:

```json
"parserOptions": {
   "ecmaVersion": 2021
}
```

The core and most of the rules have been updated in v7.8.0.

We left this issue open for changes that were considered to be breaking for `babel-eslint` users, and those changes will be implemented in ESLint v8. There are also some minor changes that were blocked at the moment, we'll implement them in some of the future v7 minor releases.

**Comment by aduh95** [comment]
> ESLint also supports logical assignment operators, as of [v7.8.0](https://eslint.org/blog/2020/08/eslint-v7.8.0-released).

That doesn't seem to be true, for the following input: `var foo ||= bar;`, I'm getting the following error: `1:9 - Parsing error: Unexpected token ||=`.

[Demo link](https://eslint.org/demo#eyJ0ZXh0IjoidmFyIGZvbyB8fD0gYmFyOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWVsc2UtaWYiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5IjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWltcG9ydC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1vYmotY2FsbHMiOjIsIm5vLW9jdGFsIjoyLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOjIsIm5vLXJlZGVjbGFyZSI6Miwibm8tcmVnZXgtc3BhY2VzIjoyLCJuby1zZWxmLWFzc2lnbiI6Miwibm8tc2V0dGVyLXJldHVybiI6Miwibm8tc2hhZG93LXJlc3RyaWN0ZWQtbmFtZXMiOjIsIm5vLXNwYXJzZS1hcnJheXMiOjIsIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjoyLCJuby11bmRlZiI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1jYXRjaCI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsIm5vLXdpdGgiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7ImVzMjAyMSI6dHJ1ZX19fQ==)

**Comment by mdjermanovic** [comment]
> var foo ||= bar;

@aduh95 this isn't valid [_VariableStatement_](https://tc39.es/ecma262/#prod-VariableStatement), `||= bar` cannot be _Initializer_. This code is syntax error in engines, too.

**Comment by TimMoser92** [comment]
I just stumbled upon a complaint from ESLint that might be related to these changes. The rule is [`no-return-assign`](https://eslint.org/docs/rules/no-return-assign) with the option `always` set (though that is not relevant). The code looks something like this follows:
```js
const resultMap = {};
const accessGroup = (group) => resultMap[group] ??= [];
```

ESlint complains here, because the return of the arrow function conatins an assignment. That is technically correct, but I do not think that it is a problem. Let's leave aside, whether or not this is the most beautiful solution. The `no-return-assign` rule is justified as follows:
> It is difficult to tell the intent of the return statement here. It's possible that the function is meant to return the result of bar + 2, but then why is it assigning to foo? It's also possible that the intent was to use a comparison operator such as == and that this code is an error.

This justification does not work as well for logical assignment operators. Since:
a) A logical assignment is not easily mistakenly written when the comparison operator `==` was intended.
b) It is a valid use-case to conditionally assign and return something, especially in an arrow function, as seen above.

The complaint even persists when the intent is shifted and thus explicated further by adding additional code (as long as the `always` option is specified):
```js
const resultMap = {};
const addElementToGroup = ({ element, group }) => (resultMap[group] ??= []).push(element);
```

My current intuition would be to exempt logical assignments from this rule. Otherwise the rule's justification might have to be extended to cover logical assignments as different cases. What do you think?

**Comment by ljharb** [comment]
@TimMoser92 the rule helps prevent conflating assignments with expressions, and it absolutely should apply to the logical assignment operators. If the docs imply different, I think they should be updated.

**Comment by TimMoser92** [comment]
> @TimMoser92 the rule helps prevent conflating assignments with expressions, and it absolutely should apply to the logical assignment operators. If the docs imply different, I think they should be updated.

If that is the correct justification, the docs do not seem to imply it. They primarily mention other specific problems that can arise from mixing non-logical assignments into return statements. The justification you give is much more general and would of course also cover logical assignment operators. But if the goal is to never mix assignments into expressions, why is there no general rule about it? Statements like `(map[key] ??= []).push(sth)` could be forbidden in general, if that were the goal.

So I am not sure, what the solution is here and where this should be decided, nor do I have a strong opinion in any direction, but it seems documented justification and behaviour do not currently match.

**Comment by mdjermanovic** [comment]
All planned tasks are now completed, so I'm closing this issue.

@TimMoser92 the behavior for `??=` is consistent with the behavior for other operators, for example:

```js
/* eslint no-return-assign: ["error", "always"] */

const accessGroup1 = (group) => resultMap[group] ??= []; // reports error

const accessGroup2 = (group) => resultMap[group] += []; // also reports error

```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJldHVybi1hc3NpZ246IFtcImVycm9yXCIsIFwiYWx3YXlzXCJdICovXG5cbmNvbnN0IGFjY2Vzc0dyb3VwMSA9IChncm91cCkgPT4gcmVzdWx0TWFwW2dyb3VwXSA/Pz0gW107IC8vIHJlcG9ydHMgZXJyb3JcblxuY29uc3QgYWNjZXNzR3JvdXAyID0gKGdyb3VwKSA9PiByZXN1bHRNYXBbZ3JvdXBdICs9IFtdOyAvLyBhbHNvIHJlcG9ydHMgZXJyb3JcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

If there seems to be something in the rule and/or the documentation for this rule that should be corrected in this regard, feel free to open a separate issue.
